### ISSUE TITLE
[BUG]: Missing typing_extensions dependency under python 3.11+

### ISSUE THREAD

**Issue by asnare** [databrickslabs/ucx#1875]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Uninstalling UCX on Windows with Python 3.12 fails due to `ModuleNotFoundError` for `typing_extenstions`.
```
C:\Users\shindea>databricks labs uninstall ucx
Traceback (most recent call last):
  File "C:\Users\shindea\.databricks\labs\ucx\lib\src\databricks\labs\ucx\uninstall.py", line 6, in <module>
    from databricks.labs.ucx.install import WorkspaceInstallation
  File "C:\Users\shindea\.databricks\labs\ucx\1ib\src\databricks\labs\ucx\install.py", line 50, in <module>
    from databricks.labs.ucx.contexts.account_cli import AccountContext
  File "C:\Users\shindea\.databricks\labs\ucx\1ib|src\databricks\labs\ucx\contexts\account_cli.py", line 7, in <module>
    from databricks.labs.ucx.account.aggregate import AccountAggregate
  File "C:\Users\shindea\.databricks\labs\ucx\lib|src\databricks\labs\ucx\account\aggregate.py", line 14, in <module>
    from databricks.labs.ucx.contexts.workspace_cli import WorkspaceContext
  File "C:\Users\shindea\.databricks\labs\ucx\lib\src\databricks\labs\ucx\contexts\workspace_cli.py", line 17, in <module>
    from databricks.labs.ucx.contexts.application import CliContext
  File "C:\Users\shindea\.databricks\labs\ucx\1ib\src\databricks\labs\ucx\contexts\application.py", line 18, in <module>
    from databricks.labs.ucx.source._code. python_libraries import PythonLibraryResolver
  File "C:\Users\shindea\.databricks\labs\ucx\lib\src\databricks\labs\ucx\source_code\python_libraries.py", line 14, in <module>
    from databricks.labs.ucx.source_code.graph import (
  File "C:\Users\shindea\.databricks\labs\ucx\lib\src\databricks\labs\ucx\source_code\graph.py", line 10, in <module>
    from databricks.labs.ucx.source_code.base import Advisory
  File "C:\Users\shindea\.databricks\labs\ucx\lib|src\databricks\labs\ucx\source_code\base.py", line 9, in <module>
    from typing extensions import Self
ModuleNotFoundError: No module named 'typing_extensions'
Error: uninstall hook: exit status 1
```

### Expected Behavior

It should not fail with this import error.

### Cloud

Azure

### Operating System

Windows

### Version

latest via Databricks CLI

### Relevant log output

```shell
C:\Users\shindea>databricks labs uninstall ucx
Traceback (most recent call last):
  File "C:\Users\shindea\.databricks\labs\ucx\lib\src\databricks\labs\ucx\uninstall.py", line 6, in <module>
    from databricks.labs.ucx.install import WorkspaceInstallation
  File "C:\Users\shindea\.databricks\labs\ucx\1ib\src\databricks\labs\ucx\install.py", line 50, in <module>
    from databricks.labs.ucx.contexts.account_cli import AccountContext
  File "C:\Users\shindea\.databricks\labs\ucx\1ib|src\databricks\labs\ucx\contexts\account_cli.py", line 7, in <module>
    from databricks.labs.ucx.account.aggregate import AccountAggregate
  File "C:\Users\shindea\.databricks\labs\ucx\lib|src\databricks\labs\ucx\account\aggregate.py", line 14, in <module>
    from databricks.labs.ucx.contexts.workspace_cli import WorkspaceContext
  File "C:\Users\shindea\.databricks\labs\ucx\lib\src\databricks\labs\ucx\contexts\workspace_cli.py", line 17, in <module>
    from databricks.labs.ucx.contexts.application import CliContext
  File "C:\Users\shindea\.databricks\labs\ucx\1ib\src\databricks\labs\ucx\contexts\application.py", line 18, in <module>
    from databricks.labs.ucx.source._code. python_libraries import PythonLibraryResolver
  File "C:\Users\shindea\.databricks\labs\ucx\lib\src\databricks\labs\ucx\source_code\python_libraries.py", line 14, in <module>
    from databricks.labs.ucx.source_code.graph import (
  File "C:\Users\shindea\.databricks\labs\ucx\lib\src\databricks\labs\ucx\source_code\graph.py", line 10, in <module>
    from databricks.labs.ucx.source_code.base import Advisory
  File "C:\Users\shindea\.databricks\labs\ucx\lib|src\databricks\labs\ucx\source_code\base.py", line 9, in <module>
    from typing extensions import Self
ModuleNotFoundError: No module named 'typing_extensions'
Error: uninstall hook: exit status 1
```

**Comment by asnare** [comment]
This bug was introduced in #1855; the `typing_extensions` module is a transitive dependency of `astroid`, but only on Python <3.11.

Proposed fix:
 1. List the `typing_extensions` module explicitly for Python <3.11.
 2. Conditionally import `Self` from either `typing` or `typing_extensions` depending on the version; this is the standard idiom.

We could also put the import behind a `TYPE_CHECKING` guard but we prefer to avoid that approach in this project.

**Comment by gangvijay** [comment]
I am facing same issue and could you please provide the detailed steps to resolve the issue.

Python Version: 3.11.1
Databricks CLI v0.212.4

**Comment by asnare** [comment]
> I am facing same issue and could you please provide the detailed steps to resolve the issue.
> 
> Python Version: 3.11.1 Databricks CLI v0.212.4

@gangvijay: The next release will resolve the issue; in the meantime the easiest fix is to use python 3.10.
